File: src/codegen/subcollection/subresource.py
#
# Copyright (c) 2010 Red Hat, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#           http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#


from codegen.utils.typeutil import TypeUtil
from codegen.collection.resource import Resource
from codegen.doc.documentation import Documentation
from codegen.utils.paramutils import ParamUtils
from ovirtsdk.utils.parsehelper import ParseHelper
from codegen.utils.urlutils import UrlUtils
from codegen.utils.headerutils import HeaderUtils


#============================================================
#=====================SUB COLLECTION RESOURCE================
#============================================================

class SubResource(object):
    @staticmethod
    def resource(sub_res_type, encapsulating_entity, parent, KNOWN_WRAPPER_TYPES={}):
        actual_xml_entity = TypeUtil.getValueByKeyOrNone(encapsulating_entity.lower(), KNOWN_WRAPPER_TYPES)
        actual_xml_type = ParseHelper.getXmlType(encapsulating_entity)
        actual_sub_res_type = TypeUtil.getValueByKeyOrNone(sub_res_type.lower(), KNOWN_WRAPPER_TYPES)

        sub_collection_resource_template_values = {'encapsulating_entity':actual_sub_res_type if actual_sub_res_type is not None else sub_res_type,
                                                   'sub_res_type': actual_sub_res_type if actual_sub_res_type is not None else \
                                                                                       actual_xml_entity if actual_xml_entity is not None else\
                                                                                       actual_xml_type.__name__ if \
                                                                                       actual_xml_type is not None else \
                                                                                       encapsulating_entity,
                                                   'parent':parent.lower(),
                                                   'encapsulated_entity':encapsulating_entity.lower()}

        sub_resource_template = \
        (
        "class %(encapsulating_entity)s(params.%(sub_res_type)s, Base):\n" + \
        "    def __init__(self, %(parent)s, %(encapsulated_entity)s, context):\n" + \
        "        Base.__init__(self, context)\n"
        "        self.parentclass = %(parent)s\n" + \
        "        self.superclass  =  %(encapsulated_entity)s\n\n" + \
        Resource.SUB_COLLECTIONS_FIXME + "\n" + \

        "    def __new__(cls, %(parent)s, %(encapsulated_entity)s, context):\n" + \
        "        if %(encapsulated_entity)s is None: return None\n" + \
        "        obj = object.__new__(cls)\n" + \
        "        obj.__init__(%(parent)s, %(encapsulated_entity)s, context)\n" + \
        "        return obj\n\n" + \

        "    def __getProxy(self):\n" + \
        "        proxy = context.manager[self.context].get('proxy')\n" + \
        "        if proxy:\n" + \
        "            return proxy\n" + \
        "        #This may happen only if sdk was explicitly disconnected\n" + \
        "        #using .disconnect() method, but resource instance ref. is\n" + \
        "        #still available at client's code.\n" + \
        "        raise DisconnectedError\n\n"
        ) % sub_collection_resource_template_values

        return sub_resource_template

    @staticmethod
    def action(url, link, action_name, parent_resource_name_lc, body_type, resource_name_lc, method, action_params={}, collection_action=False):
        sub_collection_resource_action_template_values = {'url':url,
                                                          'action_name':action_name,
                                                          'method': method,
                                                          'parent_resource_name_lc':parent_resource_name_lc.lower(),
                                                          'body_type':body_type,
                                                          'body_type_lc':body_type.lower(),
                                                          'resource_name_lc':resource_name_lc.lower(),
                                                          'add_method_params' : Resource._addMethodParams(action_params.keys()),
                                                          'add_action_parans' : Resource._addActionParams(action_params)}
        headers_method_params_str, headers_map_params_str = HeaderUtils.generate_method_params(link)
        headers_method_params_str = ', ' + headers_method_params_str if headers_method_params_str != '' else headers_method_params_str

        if collection_action == True:
            sub_collection_resource_action_template = \
            ("    def %(action_name)s(self%(add_method_params)s, %(body_type_lc)s=params.%(body_type)s()" + headers_method_params_str + "):\n" + \
            Documentation.document(link) +
            "        url = '%(url)s'\n\n" + \
            "        result = self.__getProxy().request(method='%(method)s',\n" + \
            "                                           url=UrlHelper.replace(url, {'{%(parent_resource_name_lc)s:id}' : self.parentclass.get_id()}),\n" + \
# TODO: support action on sub-collection
            "                                           body=ParseHelper.toXml(%(body_type_lc)s),\n"
            "                                           headers=" + headers_map_params_str + ")\n\n"
            "        return result\n\n"
            ) % sub_collection_resource_action_template_values
        else:
            sub_collection_resource_action_template = \
            ("    def %(action_name)s(self%(add_method_params)s, %(body_type_lc)s=params.%(body_type)s()" + headers_method_params_str + "):\n" + \
            Documentation.document(link) +
            "        url = '%(url)s'\n\n" + \
            "        result = self.__getProxy().request(method='%(method)s',\n" + \
			"                                           url=UrlHelper.replace(url, " + \
            UrlUtils.generate_url_identifiers_replacments(link,
                                                           offset="                                                                      ",
                                                           continues=True) + \
            "),\n" + \
            "                                           body=ParseHelper.toXml(%(body_type_lc)s),\n"
            "                                           headers=" + headers_map_params_str + ")\n\n"
            "        return result\n\n"
            ) % sub_collection_resource_action_template_values

#            "                                                                {'{%(parent_resource_name_lc)s:id}' : self.parentclass.get_id(),\n" + \
#            "                                                                 '{%(resource_name_lc)s:id}': self.get_id()}),\n" + \
        return sub_collection_resource_action_template

    @staticmethod
    def update(url, link, parent_resource_name_lc, resource_name, returned_type, KNOWN_WRAPPER_TYPES):
        actual_xml_entity = TypeUtil.getValueByKeyOrNone(returned_type.lower(), KNOWN_WRAPPER_TYPES)
        sub_collection_resource_update_template_values = \
        {
         'url':url,
        'parent_resource_name_lc':parent_resource_name_lc.lower(),
        'resource_name':resource_name,
        'resource_name_lc':resource_name.lower(),
        'returned_type':actual_xml_entity if actual_xml_entity is not None
                                          else returned_type
        }
        headers_method_params_str, headers_map_params_str = HeaderUtils.generate_method_params(link)
        headers_method_params_str = \
            ', ' + headers_method_params_str if headers_method_params_str != '' else headers_method_params_str
        prms_str, method_params, url_params = ParamUtils.getMethodParamsByUrlParamsMeta(link)

        if prms_str != '' or headers_method_params_str != '':
            combined_method_params = ', ' if prms_str != '' and not prms_str.startswith(',') else ''
            combined_method_params += prms_str + headers_method_params_str

        method_params_copy = method_params.copy()
        url_params = ParamUtils.toDictStr(url_params.keys(), method_params_copy.keys())
        url_identifiers_replacments = \
            UrlUtils.generate_url_identifiers_replacments(link,
                                                          offset="                                     ",
                                                          continues=True)

        sub_collection_resource_update_template = \
        ("    def update(self" + combined_method_params + "):\n" + \
         Documentation.document(link, {}, method_params) +
        "        url = '%(url)s'\n" + \
        "        url = UrlHelper.replace(url, " + url_identifiers_replacments + ")\n\n" + \
        "        result = self.__getProxy().update(url=SearchHelper.appendQuery(url" + \
        ((", " + url_params) if url_params != '' else '') + "),\n" + \
        "                                          body=ParseHelper.toXml(self.superclass),\n" + \
        "                                          headers=" + headers_map_params_str + ")\n\n" + \
        "        return %(returned_type)s(self.parentclass, result, self.context)\n\n") % sub_collection_resource_update_template_values


        return sub_collection_resource_update_template

    @staticmethod
    def delete(url, link, parent_resource_name_lc, resource_name_lc, body_type):
        sub_collection_resource_delete_template_values = {'url':url,
                                                          'body_type_lc':body_type.lower() if body_type is not None
                                                                                           else None,
                                                          'parent_resource_name_lc':parent_resource_name_lc.lower(),
                                                          'resource_name_lc':resource_name_lc.lower()}

        prms_str, method_params, url_params = ParamUtils.getMethodParamsByUrlParamsMeta(link)
        headers_method_params_str, headers_map_params_str = HeaderUtils.generate_method_params(link)
        combined_method_params = prms_str + \
                                 (', ' if prms_str != '' and headers_method_params_str != '' else '') + \
                                 headers_method_params_str
        headers_map_params_str_with_no_ct = headers_map_params_str.replace('}',
                                                                          ',"Content-type":None}') \
                                                                 if headers_map_params_str != '{}' \
                                                                 else '{"Content-type":None}'

        body_instance = ParamUtils.getBodyInstance(link)
        body_instance_str = '=' + body_instance if body_instance else ''

        if prms_str != '' or headers_method_params_str != '':
            sub_collection_resource_delete_template = \
            ("    def delete(self, " + combined_method_params + "):\n" + \
             Documentation.document(link, {}, method_params) +
             "        url = UrlHelper.replace('%(url)s',\n" + \
            UrlUtils.generate_url_identifiers_replacments(link,
                                                           offset="                                ") + \
             ")\n\n" + \
             "        return self.__getProxy().delete(url=SearchHelper.appendQuery(url, " + ParamUtils.toDictStr(url_params.keys(),
                                                                                                                method_params.copy().keys()) +
                                                                                    "),\n" +
             "                                        headers=" + headers_map_params_str_with_no_ct + ")\n\n"
             ) % sub_collection_resource_delete_template_values

            body_sub_collection_resource_delete_template = \
            ("    def delete(self, %(body_type_lc)s" + body_instance_str + ", " + combined_method_params + "):\n" + \
             Documentation.document(link, {}, method_params) +
             "        url = UrlHelper.replace('%(url)s',\n" + \
            UrlUtils.generate_url_identifiers_replacments(link,
                                                           offset="                                ") + \
             ")\n\n" + \
             "        return self.__getProxy().delete(url=SearchHelper.appendQuery(url, " + ParamUtils.toDictStr(url_params.keys(),
                                                                                                                method_params.copy().keys()) +
                                                                                    "),\n" + \
             "                                        body=ParseHelper.toXml(%(body_type_lc)s),\n"
             "                                        headers=" + headers_map_params_str + ")\n\n"
             ) % sub_collection_resource_delete_template_values
        else:
            sub_collection_resource_delete_template = \
            ("    def delete(self):\n" + \
             Documentation.document(link) +
             "        url = '%(url)s'\n\n" + \
             "        return self.__getProxy().delete(url=UrlHelper.replace(url, " + \
            UrlUtils.generate_url_identifiers_replacments(link,
                                                           offset="                                                                  ",
                                                           continues=True) + \
             "),\n" + \
             "                                       headers={'Content-type':None})\n\n"
             ) % sub_collection_resource_delete_template_values

            body_sub_collection_resource_delete_template = \
            ("    def delete(self, %(body_type_lc)s" + body_instance_str + "):\n" + \
             Documentation.document(link) +
             "        url = '%(url)s'\n\n" + \
             "        return self.__getProxy().delete(url=UrlHelper.replace(url,\n" + \
            UrlUtils.generate_url_identifiers_replacments(link,
                                                           offset="                                                             ") + \
             "),\n" + \
             "                                        body=ParseHelper.toXml(%(body_type_lc)s))\n\n"
             ) % sub_collection_resource_delete_template_values

        if not body_type:
            return sub_collection_resource_delete_template
        else:
            return body_sub_collection_resource_delete_template

File: generator/src/main/java/org/ovirt/engine/sdk/generator/python/SubResource.java
//
// Copyright (c) 2014 Red Hat, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package org.ovirt.engine.sdk.generator.python;

import java.util.LinkedHashMap;
import java.util.Map;

import org.ovirt.engine.sdk.generator.BrokerRules;
import org.ovirt.engine.sdk.generator.Location;
import org.ovirt.engine.sdk.generator.SchemaRules;
import org.ovirt.engine.sdk.generator.python.templates.SubResourceDeleteTemplate;
import org.ovirt.engine.sdk.generator.python.templates.SubResourceDeleteWithBodyTemplate;
import org.ovirt.engine.sdk.generator.python.templates.SubResourceDeleteWithUrlParamsAndBodyTemplate;
import org.ovirt.engine.sdk.generator.python.templates.SubResourceDeleteWithUrlParamsTemplate;
import org.ovirt.engine.sdk.generator.python.templates.SubResourceTemplate;
import org.ovirt.engine.sdk.generator.python.templates.SubResourceUpdateTemplate;
import org.ovirt.engine.sdk.generator.templates.AbstractTemplate;
import org.ovirt.engine.sdk.generator.python.utils.HeaderUtils;
import org.ovirt.engine.sdk.generator.python.utils.ParamUtils;
import org.ovirt.engine.sdk.generator.utils.Tree;
import org.ovirt.engine.sdk.generator.python.utils.UrlUtils;
import org.ovirt.engine.sdk.entities.DetailedLink;

public class SubResource {
    public static String resource(Tree<Location> entityTree) {
        Tree<Location> grandparentTree = entityTree.getParent().getParent();

        String entityType = SchemaRules.getSchemaType(entityTree);
        String grandparentType = SchemaRules.getSchemaType(grandparentTree);
        String brokerType = BrokerRules.getBrokerType(entityTree);
        String superClassParameterName = entityType.toLowerCase().replace("_", "");
        String parentClassParameterName = grandparentType.toLowerCase().replace("_", "");

        SubResourceTemplate template = new SubResourceTemplate();
        template.set("entity_type", entityType);
        template.set("broker_type", brokerType);
        template.set("superclass_parameter_name", superClassParameterName);
        template.set("parentclass_parameter_name", parentClassParameterName);

        return template.evaluate();
    }

    public static String update(Tree<Location> entityTree, DetailedLink link) {
        String combinedMethodParams = "";

        Object[] result = HeaderUtils.generateMethodParams(link);
        String headersMethodParamsStr = (String) result[0];
        String headersMapParamsStr = (String) result[1];

        headersMethodParamsStr = !headersMethodParamsStr.isEmpty()?
            ", " + headersMethodParamsStr:
            headersMethodParamsStr;

        result = ParamUtils.getMethodParamsByUrlParamsMeta(link);
        String prmsStr = (String) result[0];
        Map<String, String> methodParams = (Map) result[1];
        Map<String, String> urlParams = (Map) result[2];

        if (!prmsStr.isEmpty() || !headersMethodParamsStr.isEmpty()) {
            combinedMethodParams = !prmsStr.isEmpty() && !prmsStr.startsWith(",")? ", ": "";
            combinedMethodParams += prmsStr + headersMethodParamsStr;
        }

        Map<String, String> methodParamsCopy = new LinkedHashMap<>(methodParams);

        String urlParamsStr = ParamUtils.toDictStr(
            urlParams.keySet(), methodParamsCopy.keySet()
        );

        String urlIdentifiersReplacements = UrlUtils.generateUrlIdentifiersReplacements(
            entityTree,
            "            "
        );

        SubResourceUpdateTemplate template = new SubResourceUpdateTemplate();
        template.set("url", link.getHref());
        template.set("entity_broker_type", BrokerRules.getBrokerType(entityTree));
        template.set("url_identifiers_replacments", urlIdentifiersReplacements);
        template.set("headers_map_params_str", headersMapParamsStr);
        template.set("combined_method_params", combinedMethodParams);
        template.set("url_params_separator", !urlParamsStr.isEmpty()? ", " + urlParamsStr: "");
        template.set("docs", Documentation.document(link, new LinkedHashMap<>(), methodParams));

        return template.evaluate();
    }

    public static String delete(Tree<Location> collectionTree, DetailedLink link) {
        Object[] result = ParamUtils.getMethodParamsByUrlParamsMeta(link);
        String prmsStr = (String) result[0];
        LinkedHashMap<String, String> methodParams = (LinkedHashMap) result[1];
        LinkedHashMap<String, String> urlParams = (LinkedHashMap) result[2];

        result = HeaderUtils.generateMethodParams(link);
        String headersMethodParamsStr = (String) result[0];
        String headersMapParamsStr = (String) result[1];

        String combinedMethodParams =
            prmsStr +
            (!prmsStr.isEmpty() && !headersMethodParamsStr.isEmpty()? ", ": "") +
            headersMethodParamsStr;

        String headersMapParamsStrWithNoCt =
            !headersMapParamsStr.equals("{}")?
                headersMapParamsStr.replace("}", ",\"Content-type\":None}"):
                "{\"Content-type\":None}";

        String bodyInstance = ParamUtils.getBodyInstance(link);
        String bodyInstanceStr = !bodyInstance.isEmpty()? "=" + bodyInstance: "";

        String bodyType = null;
        if (link.isSetRequest() && link.getRequest().isSetBody()) {
            bodyType = link.getRequest().getBody().getType();
        }

        Map<String, String> values = new LinkedHashMap<>();
        values.put("url", link.getHref());
        values.put("body_type_lc", bodyType != null ? bodyType.toLowerCase() : null);
        values.put("headers_map_params_str_with_no_ct", headersMapParamsStrWithNoCt);
        values.put("combined_method_params", combinedMethodParams);
        values.put("body_instance_str", bodyInstanceStr);
        values.put("headers_map_params_str", headersMapParamsStr);
        values.put("url_params", ParamUtils.toDictStr(urlParams.keySet(), methodParams.keySet()));

        AbstractTemplate templateWithBody;
        AbstractTemplate templateWithoutBody;
        if (!prmsStr.isEmpty() || !headersMethodParamsStr.isEmpty()) {
            values.put("docs", Documentation.document(link, new LinkedHashMap<>(), methodParams));
            values.put("url_identifyiers",
                UrlUtils.generateUrlIdentifiersReplacements(
                    collectionTree,
                    "            "
                )
            );

            templateWithBody = new SubResourceDeleteWithUrlParamsAndBodyTemplate();
            templateWithoutBody = new SubResourceDeleteWithUrlParamsTemplate();
        }
        else {
            values.put("docs", Documentation.document(link));
            values.put("url_identifyiers",
                UrlUtils.generateUrlIdentifiersReplacements(
                    collectionTree,
                    "                "
                )
            );
            values.put("url_identifyiers_with_body",
                UrlUtils.generateUrlIdentifiersReplacements(
                    collectionTree,
                    "                                                               "
                )
            );

            templateWithBody = new SubResourceDeleteWithBodyTemplate();
            templateWithoutBody = new SubResourceDeleteTemplate();
        }

        if (bodyType == null || bodyType.isEmpty()) {
            templateWithoutBody.set(values);
            return templateWithoutBody.evaluate();
        }
        else {
            templateWithBody.set(values);
            return templateWithBody.evaluate();
        }
    }
}
